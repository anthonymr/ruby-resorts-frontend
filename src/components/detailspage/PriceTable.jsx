import { useSelector } from 'react-redux';
import { styled } from '@mui/material/styles';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, { tableCellClasses } from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableRow from '@mui/material/TableRow';
import { detailsTable } from './styles';

const StyledTableCell = styled(TableCell)(() => ({
  [`&.${tableCellClasses.body}`]: {
    fontSize: 16,
  },
}));

const StyledTableRow = styled(TableRow)(({ theme }) => ({
  '&:nth-of-type(odd)': {
    backgroundColor: theme.palette.action.hover,
  },
  ' td, th': {
    border: 0,
  },
}));

const PriceTable = () => {
  const { roomDetails } = useSelector((state) => state.details);

  function createData(name, value) {
    return { name, value };
  }

  const reservationFee = parseFloat(roomDetails.reservation_fee);
  const reservationPrice = parseFloat(roomDetails.reservation_price);

  const totalAmount = Number.isNaN(reservationFee)
    || Number.isNaN(reservationPrice)
    ? 0
    : reservationFee + reservationPrice;

  const rows = [
    createData('Reservation Price', roomDetails.reservation_price),
    createData('Transaction fee', roomDetails.reservation_fee),
    createData('Total amount', totalAmount),
    createData('Cancelation Fee', '30%'),
  ];

  return (
    <TableContainer>
      <Table sx={detailsTable}>
        <TableBody>
          {rows.map((row) => (
            <StyledTableRow key={row.name}>
              <StyledTableCell component="th" scope="row">
                {row.name}
              </StyledTableCell>
              <StyledTableCell align="right">{row.value}</StyledTableCell>
            </StyledTableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default PriceTable;
