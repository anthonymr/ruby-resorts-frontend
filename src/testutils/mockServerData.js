const userInfoData = {
  id: 1,
  name: 'Test User',
  username: 'testuser',
  email: 'testuser@test.com',
  role: 'admin',
  created_at: '2023-06-17T12:20:06.319Z',
  updated_at: '2023-06-17T12:20:47.606Z',
};

export const roomsData = [
  {
    id: 1,
    name: 'Server Test Suite 1',
    description: 'description of server test suite 1',
    image_url: 'someimageurl1',
  },
  {
    id: 2,
    name: 'Server Test Suite 2',
    description: 'description of server test suite 2',
    image_url: 'someimageurl2',
  },
];

export const hotelsData = [
  {
    id: 1,
    name: 'Test Hotel 1',
    city: 'Test City 1',
    created_at: '2023-06-17T12:15:02.781Z',
    updated_at: '2023-06-17T12:15:02.781Z',
  },
];

export const detailsData = {
  id: 1,
  name: 'Server Test Suite 1',
  description: 'Description of test suite 1',
  full_price: '100.0',
  reservation_price: '10.0',
  reservation_fee: '10.0',
  rating: 5,
  created_at: '2023-06-17T12:15:02.876Z',
  updated_at: '2023-06-17T12:15:02.887Z',
  image_url: 'some image url',
};

export const newRoomData = {
  id: 1,
  name: 'Server New Test Suite',
  description: 'Description of new test suite',
  full_price: '250.0',
  reservation_price: '200.0',
  reservation_fee: '10.0',
  rating: 5,
  created_at: '2023-06-17T12:15:02.876Z',
  updated_at: '2023-06-17T12:15:02.887Z',
  image_url: 'some image url',
};

export const reservationsData = [
  {
    id: 1,
    start_date: '2023-06-18',
    end_date: '2023-06-20',
    user_id: 1,
    hotel_id: 1,
    room_id: 1,
    created_at: '2023-06-18T07:11:15.304Z',
    updated_at: '2023-06-18T07:11:15.304Z',
    amount: '20.0',
    hotel: {
      id: 1,
      name: 'Test Hotel',
      city: 'Test City',
    },
    room: {
      id: 1,
      name: 'Test Suite',
      description:
        'description of test suite 1',
      full_price: '100.0',
      reservation_price: '10.0',
      reservation_fee: '10.0',
      rating: 5,
    },
    user: {
      id: 1,
      name: 'Server Test User 1',
      username: 'testuser',
      email: 'testuser@test.com',
      role: 'admin',
    },
  },
];

export default userInfoData;
