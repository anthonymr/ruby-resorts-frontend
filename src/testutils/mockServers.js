/*eslint-disable*/
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import userInfoData, {
  roomsData,
  hotelsData,
  detailsData,
  reservationsData,
  newRoomData,
} from './mockServerData';

const roomsHandler = [
  rest.get('http://127.0.0.1:3000/api/v1/authentication', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(userInfoData), ctx.delay(10));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/rooms', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(roomsData), ctx.delay(10));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/hotels', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(hotelsData), ctx.delay(1));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/rooms/1', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(detailsData), ctx.delay(1));
  }),
  rest.get('http://127.0.0.1:3000/api/v1/reservations', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(reservationsData), ctx.delay(1));
  }),
  rest.delete('http://127.0.0.1:3000/api/v1/rooms/1', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(detailsData), ctx.delay(1));
  }),
  rest.post('http://127.0.0.1:3000/api/v1/rooms/', (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(newRoomData), ctx.delay(1));
  }),
];

export const server1 = setupServer(...roomsHandler);
