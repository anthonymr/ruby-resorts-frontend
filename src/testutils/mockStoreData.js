import { configureStore } from '@reduxjs/toolkit';
import roomsReducer from '../redux/mainPage/roomsSlice';
import detailsReducer from '../redux/detailsPage/detailsSlice';
import citiesReducer from '../redux/newReservePage/citiesSlice';
import userReducer from '../redux/login/userSlice';
import reservationsReducer from '../redux/myReservationsPage/reservationsSlice';
import { appApi } from '../services/apiService';

const initialsStateData = {
  rooms: {
    rooms: [
      {
        id: 1,
        name: 'Test Suite',
        description: 'Description of Test Suite',
        image_url: 'Some URL',
      },
    ],
    status: 'completed',
  },
  cities: {
    hotels: [
      {
        id: 1,
        name: 'Ruby Test Hotel',
        city: 'Test City',
        created_at: '2023-06-16T03:22:52.703Z',
        updated_at: '2023-06-16T03:22:52.703Z',
      },
    ],
  },
  user: {
    accessToken: null,
    userinfo: {
      id: 1,
      name: 'Admin User',
      username: 'admin',
      email: 'test@test.com',
      role: 'admin',
      created_at: '2023-06-16T03:22:52.683Z',
      updated_at: '2023-06-16T03:22:52.683Z',
    },
    error: '',
    authStatus: 'loggedin',
    dataFetched: true,
  },
  appApi: {
    queries: {},
    mutations: {},
    provided: {},
    subscriptions: {},
    config: {
      online: true,
      focused: true,
      middlewareRegistered: false,
      refetchOnFocus: false,
      refetchOnReconnect: false,
      refetchOnMountOrArgChange: false,
      keepUnusedDataFor: 60,
      reducerPath: 'appApi',
    },
  },
  reservations: {
    reservations: [
      {
        id: 1,
        start_date: '2023-06-16',
        end_date: '2023-06-17',
        user_id: 1,
        hotel_id: 1,
        room_id: 2,
        created_at: '2023-06-16T05:26:03.849Z',
        updated_at: '2023-06-16T05:26:03.849Z',
        amount: '220.0',
        hotel: {
          id: 1,
          name: 'Ruby Amsterdam',
          city: 'Amsterdam',
        },
        room: {
          id: 2,
          name: 'Emerald',
          description:
            'Emerald Suites are second only to our Ruby Suites when it comes to luxury. The suites come with 5-star amenities',
          full_price: '250.0',
          reservation_price: '220.0',
          reservation_fee: '5.0',
          rating: 5,
        },
        user: {
          id: 1,
          name: 'Admin User',
          username: 'admin',
          email: 'test@test.com',
          role: 'admin',
        },
      },
    ],
    status: 'completed',
  },
};

export const setMockStore = (preloadedState) => configureStore({
  reducer: {
    rooms: roomsReducer,
    cities: citiesReducer,
    user: userReducer,
    [appApi.reducerPath]: appApi.reducer,
    details: detailsReducer,
    reservations: reservationsReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware({
    serializableCheck: false,
  }).concat(appApi.middleware),
  preloadedState,
});

export default initialsStateData;
