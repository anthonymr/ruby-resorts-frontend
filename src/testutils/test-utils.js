import React from 'react';
import PropTypes from 'prop-types';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/store';

const render = (ui, options) => {
  const Wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;
  Wrapper.propTypes = {
    children: PropTypes.node.isRequired,
  };

  return rtlRender(ui, { wrapper: Wrapper, ...options });
};

export * from '@testing-library/react';
export { render };
