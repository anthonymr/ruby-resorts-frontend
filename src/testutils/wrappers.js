import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { PropTypes } from 'prop-types';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import initialStateData, { setMockStore } from './mockStoreData';

export function renderWithProviders(
  ui,
  {
    preloadedState = initialStateData,
    store = setMockStore(preloadedState),
    ...renderOptions
  } = {},
) {
  function Wrapper({ children }) {
    Wrapper.propTypes = {
      children: PropTypes.node.isRequired,
    };
    return <Provider store={store}>{children}</Provider>;
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const renderWithRouter = (
  ui,
  {
    preloadedState = initialStateData,
    store = setMockStore(preloadedState),
    ...renderOptions
  } = {},
) => {
  function Wrapper({ children }) {
    Wrapper.propTypes = {
      children: PropTypes.node.isRequired,
    };
    return (
      <MemoryRouter initialEntries={['/newreservepage/0']}>
        <Provider store={store}>
          <Routes>
            <Route path="/newreservepage/:roomId" element={children} />
          </Routes>
        </Provider>
      </MemoryRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
