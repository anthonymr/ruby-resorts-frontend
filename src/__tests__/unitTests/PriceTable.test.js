import { render } from '../../testutils/test-utils'; // Import the custom render function
import PriceTable from '../../components/detailspage/PriceTable';

it('renders correctly', () => {
  const { asFragment } = render(<PriceTable />);
  expect(asFragment()).toMatchSnapshot();
});
