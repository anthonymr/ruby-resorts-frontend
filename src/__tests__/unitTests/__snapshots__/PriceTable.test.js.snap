// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
  >
    <table
      class="MuiTable-root css-gh8rj1-MuiTable-root"
    >
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
            scope="row"
          >
            Reservation Price
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
          />
        </tr>
        <tr
          class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
            scope="row"
          >
            Transaction fee
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
          />
        </tr>
        <tr
          class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
            scope="row"
          >
            Total amount
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
          >
            0
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-kxwusg-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1y7quc6-MuiTableCell-root"
            scope="row"
          >
            Cancelation Fee
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-f9hqh2-MuiTableCell-root"
          >
            30%
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
